src/database/remove.ts

```diff
@@ -1,99 +1,108 @@
-              }
-              return false;
-            });
-          case NodeSize.EMPTY:
-            return true;
-          default:
-            throw new FirebaseError("Unexpected prefetch test result: " + test, { exit: 3 });
-        }
-      })
-      .then((deleted: boolean) => {
-        if (!deleted) {
-          return;
-        }
-        if (path === this.path) {
-          this.jobStack.close();
-          logger.debug("[database][long delete stack][FINAL]", this.jobStack.stats());
-        } else {
-          const parentPath = pathLib.dirname(path);
-          const prevParentPathReference = this.waitingPath.get(parentPath);
-          if (!prevParentPathReference) {
-            throw new FirebaseError(
-              `Unexpected error: parent path reference is zero for path=${path}`,
-              { exit: 3 }
-            );
-          }
-          this.waitingPath.set(parentPath, prevParentPathReference - 1);
-          if (this.waitingPath.get(parentPath) === 0) {
-            this.jobStack.add(parentPath);
-            this.waitingPath.delete(parentPath);
-          }
+  /**
+   * @return true if this path is small (i.e. can be deleted with a single request with writeSizeLimit=tiny),
+   */
+  private async deletePath(path: string): Promise<boolean> {
+    if (await this.deleteJobStack.run(() => this.remote.deletePath(path))) {
+      return Promise.resolve(true);
+    }
+    let shallowGetBatchSize = INITIAL_SHALLOW_GET_SIZE;
+    // The range of batchSize to gradually narrow down.
+    let batchSizeLow = 1;
+    let batchSizeHigh = MAX_SHALLOW_GET_SIZE + 1;
+    let batchSize = INITIAL_DELETE_BATCH_SIZE;
+    while (true) {
+      const childrenList = await this.listStack.run(() =>
+        this.remote.listPath(path, shallowGetBatchSize)
+      );
+      if (childrenList.length === 0) {
+        return Promise.resolve(false);
+      }
+      const chunks = chunkList(childrenList, batchSize);
+      let nSmallChunks = 0;
+      for (const chunk of chunks) {
+        if (await this.deleteChildren(path, chunk)) {
+          nSmallChunks += 1;
         }
-      });
+      }
+      // Narrow the batchSize range depending on whether the majority of the chunks are small.
+      if (nSmallChunks > chunks.length / 2) {
+        batchSizeLow = batchSize;
+        batchSize = Math.floor(Math.min(batchSize * 2, (batchSizeHigh + batchSize) / 2));
+      } else {
+        batchSizeHigh = batchSize;
+        batchSize = Math.floor((batchSizeLow + batchSize) / 2);
+      }
+      // Start with small number of children to learn about an appropriate batchSize.
+      if (shallowGetBatchSize * 2 <= MAX_SHALLOW_GET_SIZE) {
+        shallowGetBatchSize = shallowGetBatchSize * 2;
+      } else {
+        shallowGetBatchSize = Math.floor(MAX_SHALLOW_GET_SIZE / batchSize) * batchSize;
+      }
+    }
+  }
+
+  private async deleteChildren(path: string, children: string[]): Promise<boolean> {
```

Comment:
This function is curious too, in the same way as the above one is. What does this actually do? What do the return values _mean_?